packages/block-editor/src/components/block-tools/use-show-block-tools.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect } from '@wordpress/data';
+import { isUnmodifiedDefaultBlock } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../../store';
+import { useHasBlockToolbar } from '../block-toolbar/use-has-block-toolbar';
+
+/**
+ * Source of truth for which block tools are showing in the block editor.
+ *
+ * @return {Object} Object of which block tools will be shown.
+ */
+export function useShowBlockTools() {
+	const hasBlockToolbar = useHasBlockToolbar();
+
+	return useSelect(
+		( select ) => {
+			const {
+				getSelectedBlockClientId,
+				getFirstMultiSelectedBlockClientId,
+				getBlock,
+				getSettings,
+				hasMultiSelection,
+				__unstableGetEditorMode,
+				isTyping,
+			} = select( blockEditorStore );
+
+			const clientId =
+				getSelectedBlockClientId() ||
+				getFirstMultiSelectedBlockClientId();
+
+			const { name = '', attributes = {} } = getBlock( clientId ) || {};
+			const editorMode = __unstableGetEditorMode();
+			const hasSelectedBlock = clientId && name;
+			const isEmptyDefaultBlock = isUnmodifiedDefaultBlock( {
+				name,
+				attributes,
+			} );
+			const _showEmptyBlockSideInserter =
+				clientId &&
+				! isTyping() &&
+				editorMode === 'edit' &&
+				isUnmodifiedDefaultBlock( { name, attributes } );
+			const maybeShowBreadcrumb =
+				hasSelectedBlock &&
+				! hasMultiSelection() &&
+				( editorMode === 'navigation' || editorMode === 'zoom-out' );
+
+			return {
+				showEmptyBlockSideInserter: _showEmptyBlockSideInserter,
+				showBreadcrumb:
+					! _showEmptyBlockSideInserter && maybeShowBreadcrumb,
+				showBlockToolbarPopover:
+					hasBlockToolbar &&
+					! getSettings().hasFixedToolbar &&
+					! _showEmptyBlockSideInserter &&
+					hasSelectedBlock &&
+					! isEmptyDefaultBlock &&
+					! maybeShowBreadcrumb,
+				showFixedToolbar:
+					editorMode !== 'zoom-out' &&
+					hasBlockToolbar &&
+					getSettings().hasFixedToolbar,
```

Comment:
I had concerns about that too. However, the information we need for `showFixedToolbar` is all right here, and this hook is already running for `<BlockTools />`. It doesn't need to do any additional work to set this value. I think it will end up being  more performant and less code, so it's worth the tradeoff.

Reply:
This is probably fine. We can always refactor if [CodeVitals](https://www.codevitals.run/project/gutenberg) catches any perf regression.

Reply:
It's fine, it's not worth trying to optimise performance here because there's only one block toolbar visible at any point in time.

Reply:
Or maybe not, looking at the comments below. Not sure if this is the cause of the regression though.